File: realtime/handlers.js


function raven_handlers(socket) {

    socket.on("fire", () => {
        socket.emit("ice");
    });

    socket.on("raven_channel_get_typers", function (channel) {
        socket.has_permission("Raven Channel", channel).then(() => {
            // Show who's typing in the channel - only send this to the user who requested it
            // User emits this event when they open the channel
            notify_typing({ socket, channel, toUser: true });

        });
    });

    socket.on("raven_channel_typing", function (channel) {

        // Join the typing room and notify users in the channel room that the user is typing
        socket.join(channel_typing_room(channel));

        notify_typing({ socket, channel, toUser: false });

    })

    socket.on("raven_channel_typing_stopped", function (channel) {


        // Leave the typing room and notify users in the channel room that the user has stopped typing

        socket.leave(channel_typing_room(channel));

        notify_typing({ socket, channel, toUser: false });

    })

}

function notify_typing(args) {
    if (!(args && args.socket && args.channel)) return;

    const socket = args.socket;
    const channel = args.channel;

    const typers_room = channel_typing_room(channel);

    const clients = Array.from(socket.nsp.adapter.rooms.get(typers_room) || []);

    let users = [];

    socket.nsp.sockets.forEach((sock) => {
        if (clients.includes(sock.id)) {
            users.push(sock.user);
        }
    });

    const channel_room = args.toUser ? user_room(args.socket.user) : open_doc_room("Raven Channel", channel);

    // notify
    socket.nsp.to(channel_room).emit("raven_channel_typers", {
        channel,
        users: Array.from(new Set(users)),
    });
}

const channel_typing_room = (channel) => "raven_channel_typing:" + channel;
const open_doc_room = (doctype, docname) => "open_doc:" + doctype + "/" + docname;
const user_room = (user) => "user:" + user;

module.exports = raven_handlers

File: raven/realtime/handlers.py
# Copyright (c) 2026, Frappe Technologies Pvt. Ltd. and contributors
# License: MIT. See LICENSE

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
	from frappe.realtime import Socket

from frappe.realtime import get_user_room, realtime


@realtime.on("fire", allow_guest=True)
def fire(socket: Socket) -> None:
	socket.emit("ice")


@realtime.on("raven_channel_get_typers")
def raven_channel_get_typers(socket: Socket, channel: str) -> None:
	# Sent when a user opens a channel: reply to that user only with who is typing.
	if socket.has_permission("Raven Channel", channel):
		notify_typing(socket, channel, to_user=True)


@realtime.on("raven_channel_typing")
def raven_channel_typing(socket: Socket, channel: str) -> None:
	# Join the typing room and tell the channel this user is typing.
	socket.join(get_channel_typing_room(channel))
	notify_typing(socket, channel, to_user=False)


@realtime.on("raven_channel_typing_stopped")
def raven_channel_typing_stopped(socket: Socket, channel: str) -> None:
	# Leave the typing room and tell the channel this user stopped.
	socket.leave(get_channel_typing_room(channel))
	notify_typing(socket, channel, to_user=False)


def current_typers(socket: Socket, channel: str) -> list[str]:
	"""Distinct users currently in the channel's typing room."""
	sids = socket.participants(get_channel_typing_room(channel))
	users = [socket.user_of(sid) for sid in sids]
	return list(dict.fromkeys(u for u in users if u))


def notify_typing(socket: Socket, channel: str, to_user: bool) -> None:
	"""Emit the current typers for a channel.

	to_user=True  -> reply only to the requesting user (their user room).
	to_user=False -> broadcast to everyone viewing the channel (open-doc room).
	"""
	if not channel:
		return

	if to_user:
		room = get_user_room(socket.user)
	else:
		room = get_open_doc_room("Raven Channel", channel)

	users = current_typers(socket, channel)
	socket.emit("raven_channel_typers", {"channel": channel, "users": users}, room=room)


def get_channel_typing_room(channel: str) -> str:
	return f"raven_channel_typing:{channel}"


def get_open_doc_room(doctype: str, docname: str) -> str:
	return f"open_doc:{doctype}/{docname}"
